### DISCUSSION TITLE
JWT Token ( Secret key )

### DISCUSSION THREAD

**Discussion by Khaled-BH** [pocketbase/pocketbase#4511] [Q&A]
Hi , I'm wondering if there's way to set own JWT Token ( Secret key ) ?  

Also to modify the payload.

**Comment by ganigeorgiev** [comment]
The question is too vague and lack details to understand what you are asking or trying to do.

In general, if you [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/), you could check the Go or JSVM tokens related helpers:
- Go:
  - https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.3/tokens
  - https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.3/tools/security#NewJWT

- JSVM:
  - https://pocketbase.io/jsvm/modules/_tokens.html
  - https://pocketbase.io/jsvm/functions/_security.createJWT.html
  
If you need further help, please elaborate on your use case and why you need a custom JWT token (_see [The XY problem](https://xyproblem.info/)_).

**Reply by Khaled-BH** [reply]
I just want to set my own secret or reveal the secret that Pocketbase made by default, So that I can generate Tokens from my own php script by example,And will be valid to make api requests to Pocketbase with it.

**Reply by ganigeorgiev** [reply]
You can't programmatically generate a JWT for a PocketBase user or admin without using the above linked APIs or without interacting manually with the SQLite `pb_data/data.db` file (which I don't recommend unless you really have a need to).

The PocketBase auth record token secret is constructed from 2 parts:
- "Common" part - this is a random string key stored in the app settings and it is shared by all auth record tokens
- "User unique" part - this is a random string key that is unique for each auth record (stored in the related auth record table)

The token secret being constructed from 2 parts allow us:
- to not store any tokens in the database
- if we want to invalidate all auth records tokens we can simply change the "Common" part
- if we want to invalidate all tokens only for a specific user we can simply change the "User unique" part

If you want to interact with PocketBase from some other backend prefer using the HTTP admin web APIs to generate token(s) - https://pocketbase.io/docs/api-admins/ (_why fixed static API keys are not supported at the moment you can check https://github.com/pocketbase/pocketbase/issues/3118_).

Call `/api/admins/auth-with-password` only the first time to generate the token or when the token has expired because this call is usually slow due to the bcrypt hash verification.
For all other cases, try to cache the token and regularly call `/api/admins/auth-refresh` to refresh its duration.

**Reply by gokulgseven** [reply]
access the sqlite pb_data/data.db file manually 
get the recordAuthtoken form  the _param table (it has a json field that has all app level configuration including all secrets)
get the usertokenkey from the users table 
join them together to form the complete  secret 

use this secret to generate your jwt 

```
const secret = param.secret;
const secretKey = userResult.tokenKey + secret;
const payload = {
    userId: userResult.id,
    email: userResult.email,
    name: userResult.name
}
const token = jwt.sign(payload, secretKey, { expiresIn: '1h' });
```

**Reply by ganigeorgiev** [reply]
@gokulgseven I don't recommend this as it is not guaranteed to work and most likely will break after the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355).

If you want a token, use the Go/JSVM helpers or the available web APIs.

**Reply by gokulgseven** [reply]
damm! i am already using it at some place .
(will keep in mind that it will stop working after the refactoring) 

till the refactoring comes, its working for me currently (without any problem since that past 3 months)

**Reply by Khaled-BH** [reply]
What I did is I've created an Admin account just for the backend like php make login request to get a token and then I normally use Admin Token to create delete etc..

My question is that a safe and good way handle these kind of requests from PHP to Pocketbase?

**Reply by ganigeorgiev** [reply]
@Khaled-BH Yes, it should be safe as long as you don't expose the generated admin token to your regular clients.
Make sure to check the foot-note in my earlier comment to avoid performance issues - https://github.com/pocketbase/pocketbase/discussions/4511#discussioncomment-8707477.

**Comment by enaaab460** [comment]
Sorry to revive this thread again, but I am trying to learn sqlite and nimlang by "recreating" core pocketbase functionality.

Right now I am trying to figure the jwt are verified, and for that I need to know where to look for the JWT secret.

+ Are passwords salted before bcrypt, if so how?
+ What is the significance of the `tokenKey` field in `users` table?
Using pocketbase 0.23

**Reply by ganigeorgiev** [reply]
@enaaab460 Please don't hijack old discussions, especially when the questions are not really related to the original one.

Regarding your questions:

1. We don't do anything special with the passwords. When using bcrypt the salt is autogenerated and stored as part of the resulting hash.
2. PocketBase JWT secret is created from 2 parts:
   - common random part stored in the specific auth collection option
   - user specific random part stored in `tokenKey` for each user
   Having the secret split this way allow us to have fully stateless APIs (no tokens are stored in the database) and at the same time we can invalidate all users tokens by just changing the common random part or if you want to invalidate the tokens for a single user - changing the `tokenKey` field (see also the explanation in https://github.com/pocketbase/pocketbase/discussions/4511#discussioncomment-8707477).

While I want to help, please note that this is not a general dev support forum so try to keep the discussions on topics related to PocketBase.
